internal/model/database.go

```diff
@@ -0,0 +1,170 @@
+package model
+
+//
+// Database
+//
+
+import (
+	"database/sql"
+	"time"
+
+	"github.com/upper/db/v4"
+)
+
+// WritabeDatabase supports writing and updating data.
+type WritableDatabase interface {
+	// Session returns the database session
+	Session() db.Session
+
+	// CreateNetwork will create a new network in the network table
+	CreateNetwork(loc LocationProvider) (*DatabaseNetwork, error)
+
+	// CreateOrUpdateURL will create a new URL entry to the urls table if it doesn't
+	// exists, otherwise it will update the category code of the one already in
+	// there.
+	CreateOrUpdateURL(urlStr string, categoryCode string, countryCode string) (int64, error)
+
+	// CreateResult writes the Result to the database a returns a pointer
+	// to the Result
+	CreateResult(homePath string, testGroupName string, networkID int64) (*DatabaseResult, error)
+
+	// UpdateUploadedStatus will check if all the measurements inside of a given result set have been uploaded and if so will set the is_uploaded flag to true
+	UpdateUploadedStatus(result *DatabaseResult) error
+
+	// Finished marks the result as done and sets the runtime
+	Finished(result *DatabaseResult) error
+
+	// DeleteResult will delete a particular result and the relative measurement on
+	// disk.
+	DeleteResult(resultID int64) error
+
+	// CreateMeasurement writes the measurement to the database a returns a pointer
+	// to the Measurement
+	CreateMeasurement(reportID sql.NullString, testName string, measurementDir string, idx int,
+		resultID int64, urlID sql.NullInt64) (*DatabaseMeasurement, error)
+
+	// AddTestKeys writes the summary to the measurement
+	AddTestKeys(msmt *DatabaseMeasurement, tk interface{}) error
+
+	// Done marks the measurement as completed
+	Done(msmt *DatabaseMeasurement) error
+
+	// UploadFailed writes the error string for the upload failure to the measurement
+	UploadFailed(msmt *DatabaseMeasurement, failure string) error
+
+	// UploadSucceeded writes the error string for the upload failure to the measurement
+	UploadSucceeded(msmt *DatabaseMeasurement) error
+
+	// Failed writes the error string to the measurement
+	Failed(msmt *DatabaseMeasurement, failure string) error
+
+	// Close closes the database session
+	Close() error
+}
+
+// ReadableDatabase only supports reading data.
+type ReadableDatabase interface {
+	// Session returns the database session
+	Session() db.Session
```

Comment:
IIUC, `Session` allows to write a database, so it should not be part of this interface.